libr/core/linux_heap_glibc.c

```diff
@@ -303,20 +334,42 @@ static void GH(print_arena_stats)(RCore *core, GHT m_arena, MallocState *main_ar
 }
 
 static bool GH(r_resolve_main_arena)(RCore *core, GHT *m_arena) {
-	if (!core || !core->dbg || !core->dbg->maps) {
-		return false;
+
+	r_return_val_if_fail (core && core->dbg && core->dbg->maps, false);
+
+	if (core->dbg->main_arena_resolved) {
+		return true;
 	}
 
 	GHT brk_start = GHT_MAX, brk_end = GHT_MAX;
 	GHT libc_addr_sta = GHT_MAX, libc_addr_end = 0;
 	GHT addr_srch = GHT_MAX, heap_sz = GHT_MAX;
+	const char *libc_path = NULL;
+	GHT libc_addr = GHT_MAX;
+	GHT main_arena_sym = GHT_MAX;
+	bool is_debugged = true;
 
 	if (r_config_get_i (core->config, "cfg.debug")) {
 		RListIter *iter;
 		RDebugMap *map;
 		r_debug_map_sync (core->dbg);
 		r_list_foreach (core->dbg->maps, iter, map) {
-			if (strstr (map->name, "/libc-") && map->perm == 6) {
+			if (strstr (map->name, "/libc-") && map->perm == R_PERM_R && main_arena_sym == GHT_MAX) {
+				libc_path = map->name;
+				libc_addr = map->addr;
+				if (!libc_path) {
+					break;
+				}
+				char *path = strdup (libc_path);
+				if (path && r_file_exists (path)) {
+					GHT vaddr = GH (get_va_symbol) (core, path, "main_arena");
+					if (libc_addr != GHT_MAX && vaddr != GHT_MAX) {
```

Comment:
probably `libc_addr != GHT_MAX` should be checked before, because if `libc_addr == GHT_MAX` we can avoid calling `r_file_exists` and `get_va_symbol`.